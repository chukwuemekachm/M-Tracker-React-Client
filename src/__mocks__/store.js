import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

export const middlewares = [thunk];
export const mockStore = configureMockStore(middlewares);
