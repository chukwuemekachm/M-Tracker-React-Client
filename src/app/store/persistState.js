/**
 * @description This module helps to persist state to the localStorage
 * As well as retrieves persisted state from the localStorage to load the app
 */

/**
 * @description Saves the state of the store to the localStorage
 *
 * @param {Object} state The state to be saved to the localStorage
 */
export const saveState = (state) => {
  const serialisedState = JSON.stringify(state);
  window.localStorage.setItem('m-tracker-react', serialisedState);
};

/**
 * @description Retrieves the state saved on the localStorage
 *
 * @returns {Object}
 */
export const loadState = () => {
  const state = window.localStorage.getItem('m-tracker-react');
  const parsedState = state !== null && state !== undefined ? JSON.parse(state) : undefined;
  if (parsedState) {
    parsedState.common.loading = false;
    parsedState.common.error = false;
  }
  return parsedState;
};

/**
 * @description Clears the state persisted on the localStorage
 */
export const clearState = () => {
  window.localStorage.removeItem('m-tracker-react');
};
