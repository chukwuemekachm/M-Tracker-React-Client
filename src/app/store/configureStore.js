import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import throttle from 'lodash/throttle';
import reducers from '../reducers';
import { loadState, saveState } from './persistState';

/* eslint-disable no-underscore-dangle */
const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
/* eslint-enable */

const preloadedState = loadState();

const store = createStore(
  reducers,
  preloadedState,
  composeEnhancers(applyMiddleware(thunk)),
);

store.subscribe(throttle(() => {
  saveState(store.getState());
}, 2000));

export default store;
