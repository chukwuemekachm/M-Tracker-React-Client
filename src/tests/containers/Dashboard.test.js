import React from 'react';
import { shallow } from 'enzyme';
import sinon from 'sinon';
import DashboardDefault, { Dashboard, mapDispatchToProps, mapStateToProps } from '../../app/containers/Dashboard';
import { mockStore } from '../../__mocks__/store';

const history = {
  push: jest.fn(),
  location: {
    search: '?pending',
  },
};
const requests = [
  {
    id: 67,
    title: 'Bad Television',
    type: 'maintenance',
    description: 'Hey this is a new request updated',
    user_id: 3,
    status: 'pending',
    createdat: '2018-09-04T18:11:56.976Z',
    updatedat: '2018-09-04T18:12:12.109Z',
  },
  {
    id: 67,
    title: 'Bad Television',
    type: 'maintenance',
    description: 'Hey this is a new request updated',
    user_id: 3,
    status: 'approved',
    createdat: '2018-09-04T18:11:56.976Z',
    updatedat: '2018-09-04T18:12:12.109Z',
  },
  {
    id: 67,
    title: 'Bad Television',
    type: 'maintenance',
    description: 'Hey this is a new request updated',
    user_id: 3,
    status: 'disapproved',
    createdat: '2018-09-04T18:11:56.976Z',
    updatedat: '2018-09-04T18:12:12.109Z',
  },
  {
    id: 67,
    title: 'Bad Television',
    type: 'maintenance',
    description: 'Hey this is a new request updated',
    user_id: 3,
    status: 'resolved',
    createdat: '2018-09-04T18:11:56.976Z',
    updatedat: '2018-09-04T18:12:12.109Z',
  },
];

const props = {
  requests,
  authenticated: true,
  getRequests: jest.fn().mockResolvedValue({ code: 200 }),
  history,
  createRequest: jest.fn().mockResolvedValue({ code: 201 }),
  filter: jest.fn(),
  filteredRequests: requests,
  getAdminRequests: jest.fn().mockResolvedValue({ code: 200 }),
  admin: true,
};
const mockState = {
  auth: {
    authenticated: true,
    user: { admin: true },
  },
  requests: {
    allRequests: requests,
    filteredRequests: requests,
  },
};
const store = mockStore(mockState);

describe('Dashboard container', () => {
  it('should render without errors', () => {
    const wrapper = shallow(<Dashboard {...props} />);
    expect(wrapper).toBeTruthy();
    expect(wrapper).toMatchSnapshot();
  });

  it('should render connected component without errors', () => {
    const wrapper = shallow(<DashboardDefault store={store} {...props} />);
    expect(wrapper.length).toBe(1);
  });

  it('should redirect when a user is not authenticated', () => {
    const wrapper = shallow(<Dashboard {...props} authenticated={false} />);
    expect(wrapper).toBeTruthy();
  });

  it('should render without errors when there are no requests', () => {
    const wrapper = shallow(<Dashboard {...props} requests={[]} filteredRequests={[]} />);
    expect(wrapper).toBeTruthy();
  });

  it('should render without errors when there are no filtered requests', () => {
    const wrapper = shallow(<Dashboard {...props} filteredRequests={[]} />);
    expect(wrapper).toBeTruthy();
  });

  it('should handle fetching all users requests', () => {
    const wrapper = shallow(<Dashboard {...props} admin={false} />);
    expect(wrapper).toBeTruthy();
  });

  it('should handle redirect when user token is expired', () => {
    const wrapper = shallow(<Dashboard
      {...props}
      admin={false}
      getRequests={jest.fn().mockResolvedValue({ code: 401 })}
    />);
    expect(wrapper).toBeTruthy();
  });

  it('should handle redirect when admin token is expired', () => {
    const wrapper = shallow(<Dashboard
      {...props}
      getAdminRequests={jest.fn().mockResolvedValue({ code: 401 })}
    />);
    expect(wrapper).toBeTruthy();
  });

  it('should mapDispatchToProps', () => {
    const sinonSpy = sinon.spy();
    const result = mapDispatchToProps(sinonSpy);
    result.getAdminRequests();
    result.getRequests();
    result.createRequest();
    result.filter();
    expect(sinonSpy.callCount).toBe(4);
  });

  it('should mapStateToProps', () => {
    expect(mapStateToProps(mockState, { history }))
      .toEqual({
        requests,
        filteredRequests: requests,
        authenticated: mockState.auth.authenticated,
        history,
        admin: mockState.auth.user.admin,
      });
  });
});
