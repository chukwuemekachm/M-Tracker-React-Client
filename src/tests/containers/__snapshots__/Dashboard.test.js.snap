// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard container should render without errors 1`] = `
<div>
  <div
    className="dashboard-body"
  >
    <Connect(NavBar) />
    <div
      className="container mt-5"
    >
      <div
        className="row"
      >
        <div
          className="col-md-3"
        >
          <SideNav
            handleNavigation={[Function]}
            requests={
              Array [
                Object {
                  "createdat": "2018-09-04T18:11:56.976Z",
                  "description": "Hey this is a new request updated",
                  "id": 67,
                  "status": "pending",
                  "title": "Bad Television",
                  "type": "maintenance",
                  "updatedat": "2018-09-04T18:12:12.109Z",
                  "user_id": 3,
                },
                Object {
                  "createdat": "2018-09-04T18:11:56.976Z",
                  "description": "Hey this is a new request updated",
                  "id": 67,
                  "status": "approved",
                  "title": "Bad Television",
                  "type": "maintenance",
                  "updatedat": "2018-09-04T18:12:12.109Z",
                  "user_id": 3,
                },
                Object {
                  "createdat": "2018-09-04T18:11:56.976Z",
                  "description": "Hey this is a new request updated",
                  "id": 67,
                  "status": "disapproved",
                  "title": "Bad Television",
                  "type": "maintenance",
                  "updatedat": "2018-09-04T18:12:12.109Z",
                  "user_id": 3,
                },
                Object {
                  "createdat": "2018-09-04T18:11:56.976Z",
                  "description": "Hey this is a new request updated",
                  "id": 67,
                  "status": "resolved",
                  "title": "Bad Television",
                  "type": "maintenance",
                  "updatedat": "2018-09-04T18:12:12.109Z",
                  "user_id": 3,
                },
              ]
            }
          />
        </div>
        <div
          className="col-md-8"
        >
          <Link
            key="67"
            replace={false}
            to="/67"
          >
            <UserRequest
              createdat="2018-09-04T18:11:56.976Z"
              description="Hey this is a new request updated"
              id={67}
              key="67"
              status="pending"
              title="Bad Television"
              type="maintenance"
              updatedat="2018-09-04T18:12:12.109Z"
              user_id={3}
            />
          </Link>
          <Link
            key="67"
            replace={false}
            to="/67"
          >
            <UserRequest
              createdat="2018-09-04T18:11:56.976Z"
              description="Hey this is a new request updated"
              id={67}
              key="67"
              status="approved"
              title="Bad Television"
              type="maintenance"
              updatedat="2018-09-04T18:12:12.109Z"
              user_id={3}
            />
          </Link>
          <Link
            key="67"
            replace={false}
            to="/67"
          >
            <UserRequest
              createdat="2018-09-04T18:11:56.976Z"
              description="Hey this is a new request updated"
              id={67}
              key="67"
              status="disapproved"
              title="Bad Television"
              type="maintenance"
              updatedat="2018-09-04T18:12:12.109Z"
              user_id={3}
            />
          </Link>
          <Link
            key="67"
            replace={false}
            to="/67"
          >
            <UserRequest
              createdat="2018-09-04T18:11:56.976Z"
              description="Hey this is a new request updated"
              id={67}
              key="67"
              status="resolved"
              title="Bad Television"
              type="maintenance"
              updatedat="2018-09-04T18:12:12.109Z"
              user_id={3}
            />
          </Link>
        </div>
      </div>
      <div
        className="col-md-1"
      />
    </div>
  </div>
  <CreateRequestForm
    create={[MockFunction]}
    history={
      Object {
        "location": Object {
          "search": "?pending",
        },
        "push": [MockFunction],
      }
    }
  />
</div>
`;
